### ISSUE TITLE
BF16 PEFT training references

### ISSUE THREAD

**Issue by heeju-kim2** [heeju-kim2/workspace_w-intern#1]
## Reference code
- Llama-recipes  code 
  [https://github.com/meta-llama/llama-recipes/tree/b7fd81c71239c67345d897c0eb6529eba076e8b8](https://github.com/meta-llama/llama-recipes/tree/b7fd81c71239c67345d897c0eb6529eba076e8b8)

**Comment by heeju-kim2** [comment]
- Revisiting BFloat16 Training (2021, sambanova)

   https://arxiv.org/pdf/2010.06192

- Doubling Neural Network Finetuning Efficiency with 16-bit Precision Techniques

   https://lightning.ai/blog/doubling-neural-network-finetuning-efficiency-with-16-bit-precision-techniques/

   https://github.com/rasbt/ViT-finetuning-scripts/tree/main

- Finetuning LLMs with LoRA and QLoRA: Insights from Hundreds of Experiments

   https://lightning.ai/pages/community/lora-insights/

- Benchmarking the Performance and Energy Efficiency of AI Accelerators for AI Training (https://arxiv.org/abs/1909.06842)

**Comment by heeju-kim2** [comment]
### Evaluation metrics

- loss per iteration
- efficiency and accuracy tradeoff
- memory cost vs accuracy
- training throughput speedup vs accuracy

**Comment by dylee-furiosa** [comment]
llama2 3runs wnadb: https://wandb.ai/debiasing/llama_recipes?nw=nwuserbig_whale

mixed : use_fp16 = False, mixed_precision = True
fp16 : use_fp16 = True, mixed_precision = False
fp32 : use_fp16 = False, mixed_precision = False

**Comment by dylee-furiosa** [comment]
`# bf16 only
output_dir="samsum_outputs/${model_name}/bf16/ep_${num_epochs}_lr_${lr}_bs_${batch_size}_wd_${weight_decay}"
mkdir -p $output_dir
torchrun --nnodes 1 --nproc_per_node 1 \
        llama-recipes/recipes/finetuning/finetuning.py \
        --enable_fsdp --use_peft --peft_method lora --dataset samsum_dataset \
        --lr $lr --num_epochs $num_epochs --batch_size_training $batch_size_training \
        --gradient_accumulation_steps $gradient_accumulation_steps --weight_decay $weight_decay \
        --model_name $model_path \
        --output_dir $output_dir \
        --pure_bf16 --optimizer anyprecision --mixed_precision False --use_fp16 False \
`

`# # bf16 + fp32 mixed precision
output_dir="samsum_outputs/${model_name}/mixed/ep_${num_epochs}_lr_${lr}_bs_${batch_size}_wd_${weight_decay}"
mkdir -p $output_dir
torchrun --nnodes 1 --nproc_per_node 1 \
        llama-recipes/recipes/finetuning/finetuning.py \
        --enable_fsdp --use_peft --peft_method lora --dataset samsum_dataset \
        --lr $lr --num_epochs $num_epochs --batch_size_training $batch_size_training \
        --gradient_accumulation_steps $gradient_accumulation_steps --weight_decay $weight_decay \
        --model_name $model_path \
        --output_dir $output_dir \
`

`
# # fp32 only
output_dir="samsum_outputs/${model_name}/fp32/ep_${num_epochs}_lr_${lr}_bs_${batch_size}_wd_${weight_decay}"
mkdir -p $output_dir
torchrun --nnodes 1 --nproc_per_node 1 \
        llama-recipes/recipes/finetuning/finetuning.py \
        --enable_fsdp --use_peft --peft_method lora --dataset samsum_dataset \
        --lr $lr --num_epochs $num_epochs --batch_size_training $batch_size_training \
        --gradient_accumulation_steps $gradient_accumulation_steps --weight_decay $weight_decay \
        --mixed_precision False --use_fp16 False \
        --model_name $model_path \
        --output_dir $output_dir \
`

**Comment by dylee-furiosa** [comment]
# bf16 only
- model parameter, hidden rep, logit 모두 bfloat16 으로 계산
- logit bflaot16 -> float32로 casting 후 loss float32로 계산

# mixed (bf16 + fp32)
- model parameter float32 -> bfloat 16으로 casting 후 hidden rep bfloat 16으로 계산
- logit bfloat 16 -> float32으로 casting 후 loss float32로 계산

# fp32 only
- model parameter, hidden rep, logit, loss 모두 float32

**Comment by dylee-furiosa** [comment]
![datatypes_page-0001](https://github.com/heeju-kim2/workspace_w-intern/assets/173657957/5c4e14f2-12db-480d-a261-5b9395cc3bde)
![datatypes_page-0002](https://github.com/heeju-kim2/workspace_w-intern/assets/173657957/f1bb0858-fd5a-4cc2-915f-d1bed2e399e2)
